src/Shared/WeakStringCache.cs

```diff
@@ -86,16 +89,24 @@ public void Free()
         /// <summary>
         /// Improvised capacity for the scavenging heuristics.
         /// </summary>
-        private int _capacity = 100;
+        private int _capacity = 101;
 
         /// <summary>
         /// A dictionary mapping string hash codes to weak GC handles.
         /// </summary>
+#if CLR2COMPATIBILITY
```

Comment:
I find ifdeffed types really hard to read and understand. I suggest making an interface or abstract class out of OpportunisticIntern with two implementations, a super simple, short, and non-optimized used just by MSBuildTaskHost, and the WeakStringCache implementation for the rest. MSBuildTaskHost.csproj would have a reference to the IOpportunisticIntern.cs interface / abstract class, and provide its own simple implementation.

Reply:
In the core libraries we use partial classes wherever possible. Sometimes with partial methods, sometimes with a private "XXXCore()" method with divergent implementations that the single entrypoint API calls into. Usually for OS specific impl but sometimes for target framework. They seem to be easier to read and of course have zero runtime impact. Possibly an option here.

Reply:
True, that's another way of separating common interface from varying implementation, as long as it actually achieves a good compromise between cognitive separation of concerns and runtime perf.

Reply:
I have split WeakStringCache.cs into three files and used partial classes instead of ifdefs. It is definitely cleaner now and should be easy for an otherwise uninitiated reader to understand. At least after they realize that only one of WeakStringCache.Locking.cs and WeakStringCache.Concurrent.cs is always included in the build. I'm wondering if there's a better way to name the files to make it more obvious.

I have also tweaked the Concurrent implementation and convinced myself that it is now free of race conditions. It is complex enough that maybe writing a small stress test would be worthwhile. Please review this code carefully.

Reply:
Looks good. Maybe change the filename to signify which is the implementation specific for the taskhost, like `WeakStringCache.Locking.cs` -> `WeakStringCache.Concurrent.MSBuildTaskHost.cs`. Maybe that makes it more clear which is the swapable logic. But even as it is, I find it pretty clear :)

Reply:
> I have also tweaked the Concurrent implementation and convinced myself that it is now free of race conditions. It is complex enough that maybe writing a small stress test would be worthwhile. Please review this code carefully.

Yea, I have a hard time reasoning whether there are any incorrect thread interleavings. On the other hand, as far as I can tell, the worst case interleaving just busts some of the weak handles with the cost of having to materialize extra IInternables right after a scavenge with the incorrect interleavings. The alternative would be to put a readers/writers lock around scavenging. Scavenging takes the writers lock, and `GetOrCreateEntry` takes the readers lock.

Reply:
>Another option to make it more clear which implementation serves which purpose is to move WeakStringCache.Locking.cs from the Shared directory to the msbuildtaskhost assembly

Great point, done!

> The alternative would be to put a readers/writers lock around scavenging. Scavenging takes the writers lock, and GetOrCreateEntry takes the readers lock.

I wanted to avoid this. Acquiring the read lock has its overhead even if there's no contention and in highly parallel scenarios (QuickBuild evaluation) scavenging would basically stop all other threads and that's not desirable.

Reply:
> The stress test would help :)

Stress test passed :)

```C#
            for (int i = 0; i < Environment.ProcessorCount; i++)
            {
                Thread thread = new Thread(() =>
                {
                    Random rnd = new Random();
                    while (true)
                    {
                        OpportunisticIntern.InternStringIfPossible(rnd.Next().ToString());
                    }
                });
                thread.Start();
            }
```

I don't think stress should be part of unit tests due to its long-running nature. I'll think about running something like this elsewhere.